lib/gossip/damper.js

```diff
@@ -0,0 +1,542 @@
+    var rVal = Math.min(config.get('dampReqRVal'), dampReqMembers.length);
+    if (rVal === 0) {
+        this.logger.warn('ringpop damper skipped subprotocol; there are not enough selectable damp req members', {
+            local: this.ringpop.whoami(),
+            flappers: flapperAddrs,
+            rVal: rVal,
+            nVal: nVal,
+            numDampReqMembers: dampReqMembers.length
+        });
+        this.emit('event', new events.DampReqUnsatisfiedEvent());
+        callback();
+        return;
+    }
+
+    this._fanoutDampReqs(flapperAddrs, dampReqMembers, rVal, onDampReqs);
+
+    function onDampReqs(err, res) {
+        if (err) {
+            self.ringpop.stat('increment', 'damper.damp-req.error');
+            self.logger.warn('ringpop damper errored out', {
+                local: self.ringpop.whoami(),
+                dampReqMembers: dampReqMemberAddrs,
+                errors: err
+            });
+            self.emit('event', new events.DampReqFailedEvent(err));
+            callback();
+            return;
+        }
+
+        self.logger.debug('ringpop damper received all damp-req responses', {
+            local: self.ringpop.whoami(),
+            flappers: flapperAddrs,
+            responses: JSON.stringify(res)
+        });
+        var membersToDamp = Damper.getMembersToDamp(res, rVal, config);
+        if (membersToDamp.length === 0) {
+            self.ringpop.stat('increment', 'damper.damp-req.inconclusive');
+            self.logger.warn('ringpop damper inconclusive', {
+                local: self.ringpop.whoami(),
+                dampReqMembers: dampReqMemberAddrs,
+                results: res
+            });
+            self.emit('event', new events.DampingInconclusiveEvent());
+            callback();
+            return;
+        }
+
+        for (var i = 0; i < membersToDamp.length; i++) {
+            self.dampMember(membersToDamp[i]);
+        }
+
+        self.ringpop.stat('increment', 'damper.damp-req.damped');
+        self.logger.warn('ringpop damped members', {
+            local: self.ringpop.whoami(),
+            dampReqMembers: dampReqMemberAddrs,
+            membersToDamp: membersToDamp,
+            results: res
+        });
+        self.emit('event', new events.DampedEvent());
+        callback();
+    }
+};
+
+Damper.prototype.removeFlapper = function removeFlapper(flapper) {
+    var address = flapper.address || flapper;
+    if (!this.flappers[address]) {
+        this.logger.debug('ringpop flapper has not been added to the damper', {
+            local: this.ringpop.whoami(),
+            flapper: address
+        });
+        return false;
+    }
+
+    delete this.flappers[address];
+    this.logger.debug('ringpop damper removed flapper', {
+        local: this.ringpop.whoami(),
+        flapper: address
+    });
+    this.ringpop.stat('increment', 'damper.flapper.removed');
+    this.ringpop.stat('gauge', 'damper.flappers',
+        Object.keys(this.flappers.length));
```

Comment:
calling `keys` on a number/undefined. You probably want `Object.keys(this.flappers).length` here.